### PR TITLE
UAR-274: Add corporate MO details page

### PR DESCRIPTION
**Pull request by bbell-companieshouse**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-274

### Change description
Implementation of a new page to facilitate the addition of a corporate managing officer.

### Work checklist

- [x] Tests added where applicable
- [x] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/config/index.ts [hunk 1]
```diff
@@ -67,6 +67,7 @@ export const UPDATE_CONFIRMATION_PAGE = "update-confirmation";
 export const OVERSEAS_ENTITY_QUERY_PAGE = "overseas-entity-query";
 export const MANAGING_OFFICER_PAGE = "managing-officer";
 export const MANAGING_OFFICER_CORPORATE_PAGE = "managing-officer-corporate";
+export const UPDATE_MANAGING_OFFICER_CORPORATE_PAGE = "update-managing-officer-corporate";
```
Comment (stephanie-ch) [inline src/config/index.ts:70]: Group this variable with the other Update ones

File: src/utils/managing.officer.corporate.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+  return res.render(templateName, {
+    backLinkUrl: backLinkUrl,
+    templateName: templateName
+  });
+};
+
+export const getManagingOfficerCorporateById = (req: Request, res: Response, next: NextFunction, templateName: string, backLinkUrl: string) => {
+  try {
+    logger.debugRequest(req, `GET BY ID ${req.route.path}`);
```
Comment (stephanie-ch) [inline src/utils/managing.officer.corporate.ts:38]: Can use req.method for the GET part of this

File: src/utils/managing.officer.corporate.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    const id = req.params[ID];
+    const data = getFromApplicationData(req, ManagingOfficerCorporateKey, id, true);
```
Comment (stephanie-ch) [inline src/utils/managing.officer.corporate.ts:41]: I changed data to officerData. You might like to do the same, but you don't need to

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+mockServiceAvailabilityMiddleware.mockImplementation((req: Request, res: Response, next: NextFunction) => next() );
+
+const mockSetApplicationData = setApplicationData as jest.Mock;
+const mockPrepareData = prepareData as jest.Mock;
+const mockLoggerDebugRequest = logger.debugRequest as jest.Mock;
+const mockMapFieldsToDataObject = mapFieldsToDataObject as jest.Mock;
+
+const DUMMY_DATA_OBJECT = { dummy: "data" };
+
+describe("MANAGING_OFFICER CORPORATE controller", () => {
```
Comment (AndreKramer7654) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:82]: add update to description

File: views/update/update-managing-officer-corporate.html [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{% extends "layout.html" %}
+
+{% set title = "Tell us about the corporate managing officer (NOT LIVE)" %}
```
Comment (AndreKramer7654) [inline views/update/update-managing-officer-corporate.html:3]: Removed (NOT LIVE) on other MO page I think. Strictly speaking this is not approved UI design so I'd keep it.

File: views/includes/page-templates/managing-officer-corporate.html [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        {% include "includes/save-and-continue-button.html" %}
+  
+        {% set removePathURL = OE_CONFIGS.MANAGING_OFFICER_CORPORATE_URL + OE_CONFIGS.REMOVE + "/" + id %}
+        {% set nameToBeRemoved = name %}
+        {% include "includes/remove-button.html" %}
+  
+      </form>
+  
+    </div>
+  </div>
```
Comment (sjahir) [inline views/includes/page-templates/managing-officer-corporate.html:78]: add newline

File: views/includes/page-templates/managing-officer-corporate.html [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        {% include "includes/inset-text.html" %}
+  
+        {% include "includes/save-and-continue-button.html" %}
+  
+        {% set removePathURL = OE_CONFIGS.MANAGING_OFFICER_CORPORATE_URL + OE_CONFIGS.REMOVE + "/" + id %}
+        {% set nameToBeRemoved = name %}
+        {% include "includes/remove-button.html" %}
+  
+      </form>
+
```
Comment (sjahir) [inline views/includes/page-templates/managing-officer-corporate.html:76]: remove empty line

File: views/includes/page-templates/managing-officer-corporate.html [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        }) }}
+  
+        {% include "includes/inset-text.html" %}
+  
+        {% include "includes/save-and-continue-button.html" %}
+  
+        {% set removePathURL = OE_CONFIGS.MANAGING_OFFICER_CORPORATE_URL + OE_CONFIGS.REMOVE + "/" + id %}
+        {% set nameToBeRemoved = name %}
+        {% include "includes/remove-button.html" %}
+
```
Comment (sjahir) [inline views/includes/page-templates/managing-officer-corporate.html:74]: remove empty line

File: views/includes/page-templates/managing-officer-corporate.html [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    <div class="govuk-grid-column-two-thirds">
+  
+      {% include "includes/list/errors.html" %}
+  
+      <span class="govuk-caption-xl">Managing officer</span>
+      <h1 class="govuk-heading-xl govuk-!-margin-0">{{ title }}</h1>
+      <br> <p class="govuk-body">You can add more later.</p> <br>
+  
+      <form class="form" method="post">
+
```
Comment (sjahir) [inline views/includes/page-templates/managing-officer-corporate.html:11]: remove empty line

File: views/includes/page-templates/managing-officer-corporate.html [hunk 10]
```diff
@@ -0,0 +1,78 @@
+<div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
```
Comment (sjahir) [inline views/includes/page-templates/managing-officer-corporate.html:2]: indentation on this file looks incorrect. It should be indented by 2 spaces

File: views/update/update-managing-officer-corporate.html (hunk 5 above)
Comment (bbell-companieshouse) [reply views/update/update-managing-officer-corporate.html:3]: Pending confirmation from BAs, but I believe it is acceptable to have for now

**Review by cdrainCH — APPROVED** [review]

File: views/includes/page-templates/managing-officer-corporate.html [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      {% include "includes/inset-text.html" %}
+
+      {% include "includes/save-and-continue-button.html" %}
+
+      {% set removePathURL = OE_CONFIGS.MANAGING_OFFICER_CORPORATE_URL + OE_CONFIGS.REMOVE + "/" + id %}
+      {% set nameToBeRemoved = name %}
+      {% include "includes/remove-button.html" %}
+    </form>
+  </div>
+</div>
```
Comment (sjahir) [inline views/includes/page-templates/managing-officer-corporate.html:75]: Add empty line

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+      mockLoggerDebugRequest.mockImplementationOnce( () => { throw new Error(ANY_MESSAGE_ERROR); });
+      const resp = await request(app).get(UPDATE_MANAGING_OFFICER_CORPORATE_URL);
+
+      expect(resp.status).toEqual(500);
+      expect(resp.text).toContain(SERVICE_UNAVAILABLE);
+    });
+  });
+
+  describe("POST tests", () => {
+
```
Comment (sjahir) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:114]: remove empty line

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+      const resp = await request(app)
+        .post(UPDATE_MANAGING_OFFICER_CORPORATE_URL)
+        .send(carriageReturnMock);
+
+      expect(resp.status).toEqual(200);
+      expect(resp.text).toContain(MANAGING_OFFICER_CORPORATE_PAGE_TITLE);
+      expect(resp.text).not.toContain(ErrorMessages.ROLES_AND_RESPONSIBILITIES_INVALID_CHARACTERS);
+    });
+
+    test(`Service address from the ${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE} is present when same address is set to no`, async () => {
```
Comment (sjahir) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:337]: add the word page after ${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE}

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+      mockPrepareData.mockImplementation( () => MANAGING_OFFICER_CORPORATE_OBJECT_MOCK_WITH_SERVICE_ADDRESS_NO);
+      await request(app)
+        .post(UPDATE_MANAGING_OFFICER_CORPORATE_URL)
+        .send(REQ_BODY_MANAGING_OFFICER_CORPORATE_MOCK_WITH_ADDRESS);
+      expect(mapFieldsToDataObject).toHaveBeenCalledWith(expect.anything(), ServiceAddressKeys, AddressKeys);
+      const data: ApplicationDataType = mockSetApplicationData.mock.calls[0][1];
+      expect(data[ServiceAddressKey]).toEqual(DUMMY_DATA_OBJECT);
+    });
+
+    test(`Service address from the ${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE} is empty when same address is set to yes`, async () => {
```
Comment (sjahir) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:347]: add the word `page` after `${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE}`

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+      mockPrepareData.mockImplementation( () => MANAGING_OFFICER_CORPORATE_OBJECT_MOCK_WITH_SERVICE_ADDRESS_YES);
+      await request(app)
+        .post(UPDATE_MANAGING_OFFICER_CORPORATE_URL)
+        .send(REQ_BODY_MANAGING_OFFICER_CORPORATE_MOCK_WITH_ADDRESS);
+      expect(mapFieldsToDataObject).not.toHaveBeenCalledWith(expect.anything(), ServiceAddressKeys, AddressKeys);
+      const data: ApplicationDataType = mockSetApplicationData.mock.calls[0][1];
+      expect(data[ServiceAddressKey]).toEqual({});
+    });
+
+    test(`Public register data from the ${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE} is present when is on register set to yes`, async () => {
```
Comment (sjahir) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:357]: add the word `page` after `${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE}`

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+      mockPrepareData.mockImplementation( () => MANAGING_OFFICER_CORPORATE_OBJECT_MOCK_WITH_PUBLIC_REGISTER_DATA_YES);
+      await request(app)
+        .post(UPDATE_MANAGING_OFFICER_CORPORATE_URL)
+        .send(REQ_BODY_MANAGING_OFFICER_CORPORATE_MOCK_WITH_ADDRESS);
+      const data: ApplicationDataType = mockSetApplicationData.mock.calls[0][1];
+      expect(data[PublicRegisterNameKey]).toEqual("Reg");
+      expect(data[RegistrationNumberKey]).toEqual("123456");
+    });
+
+    test(`Public register data from the ${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE} is empty when is on register set to no`, async () => {
```
Comment (sjahir) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:367]: add the word `page` after `${UPDATE_MANAGING_OFFICER_CORPORATE_PAGE}`

File: test/controllers/update/update.managing.officer.corporate.spec.ts [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+      mockPrepareData.mockImplementation( () => MANAGING_OFFICER_CORPORATE_OBJECT_MOCK_WITH_PUBLIC_REGISTER_DATA_NO);
+      await request(app)
+        .post(UPDATE_MANAGING_OFFICER_CORPORATE_URL)
+        .send(REQ_BODY_MANAGING_OFFICER_CORPORATE_MOCK_WITH_ADDRESS);
+      const data: ApplicationDataType = mockSetApplicationData.mock.calls[0][1];
+      expect(data[PublicRegisterNameKey]).toEqual("");
+      expect(data[RegistrationNumberKey]).toEqual("");
+    });
+
+    test("renders the next page and no errors are reported if email has leading and trailing spaces", async () => {
```
Comment (sjahir) [inline test/controllers/update/update.managing.officer.corporate.spec.ts:377]: this test description should mention the name of the next page.
`renders the next page <name of next page> and no errors...`

File: views/update/update-managing-officer-corporate.html (hunk 5 above)
Comment (bbell-companieshouse) [reply views/update/update-managing-officer-corporate.html:3]: Confirmed with Matthew Poland that having '(NOT LIVE)' is fine for now

**Review by corybudd-ch — APPROVED** [review]

File: src/utils/managing.officer.corporate.ts [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    return res.redirect(nextPage);
+  } catch (error) {
+    logger.errorRequest(req, error);
+    next(error);
+  }
+};
+
+const officerData = (reqBody: any, id: string): ApplicationDataType => {
```
Comment (AndreKramer7654) [inline src/utils/managing.officer.corporate.ts:122]: setOfficerData?

Comment (bbell-companieshouse) [reply src/utils/managing.officer.corporate.ts:122]: I believe in [Stephanies PR](https://github.com/companieshouse/overseas-entities-web/pull/684), its set to officerData so I changed it to be the same for consistency.

Comment (mwestacott) [reply src/utils/managing.officer.corporate.ts:122]: Looks like it's `setOfficerData` there.

**Comment by mwestacott** [comment]
JIRA link in original commit message too long. Ideally needs fixing.

One open comment and conflicts also need addressing.

**Review by mwestacott — APPROVED** [review]

**Pull request merged**